### ISSUE TITLE
Enable tree shaking of platform specific code

### ISSUE THREAD

**Issue by mit-mit** [flutter/flutter#14233]
Flutter apps can conditionally render platform specific UI using the APIs in [platform.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/platform.dart).

Currently these APIs depend on non-const state, resulting in this conditionally code not being tree-shakeable. We should support making this const, at least in release builds.

**Comment by mit-mit** [comment]
Related Dart SDK issue: https://github.com/dart-lang/sdk/issues/31969

**Comment by zoechi** [comment]
Seems to be related to #35705

**Comment by mkustermann** [comment]
/cc @mit-mit @lrhn  As we recently discussed this, I wanted to give an update.

@sstrickl has now plugged through the initial support.

With flutter `master` channel, one can now have code depending on the platform getters (or abstractions on top of it - if those abstractions get the `@pragma('vm:platform-const')` annotation).


Here's an example: Use `flutter create` to make a hello world app and add this:
```dart
import 'dart:io';

@pragma('vm:platform-const')
@pragma('vm:never-inline')       // To ensure platform-specific code is tree shaken even if inlining is disabled.
bool get skipLogic => Platform.isAndroid;

main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  Widget build(BuildContext context) {
  if (skipLogic) return Text('hello');
  return MaterialApp(...);
}
```

Building the app via `flutter build apk --release`:

Before the the platform-specific tree shaking support:
```
8202335  ./build/app/outputs/apk/release/app-release.apk
```
After the platform-specific tree shaking support
```
16374361 ./build_small/app/outputs/apk/release/app-release.apk
```

What remains are
* [ ] Expand the expression language we support in those abstractions.
* [ ] Make it an error/warning if a `@pragma('vm:platform-const')` was used but couldn't be evaluated at build time
* [ ] Possibly generalize the mechanism and rename to `@pragma('vm:evaluate-for-release-builds')` (or something like this)

**Comment by sstrickl** [comment]
Re: your todo list:

> Make it an error/warning if a @pragma('vm:platform-const') was used but couldn't be evaluated at build time

You should already get a failure during TFA if you can't evaluate a field initializer or method annotated with that `@pragma` to a constant. Do you have an example that doesn't? (If so, please let me know!)

**Comment by mkustermann** [comment]
> You should already get a failure during TFA if you can't evaluate a field initializer or method annotated with that @pragma to a constant. Do you have an example that doesn't? (If so, please let me know!)

I can happily compile this:
```dart
@pragma('vm:platform-const')
main() => print("hello");
```

**Comment by hunterwilhelm** [comment]
> ```dart
> @pragma('vm:platform-const')
> @pragma('vm:never-inline')       // To ensure platform-specific code is tree shaken even if inlining is disabled.
> bool get skipLogic => Platform.isAndroid;
> ```

I found that using a `final` also works for tree shaking. Thanks!

```dart
@pragma('vm:platform-const')
@pragma('vm:never-inline')
final skipLogic = Platform.isAndroid;
```

**Comment by sstrickl** [comment]
That's right, as the transformation converts any use of a static field that has an initializer or a static getter to the appropriate constant if 

* the target OS is specified,
* the static field or static getter is annotated with `@pragma('vm:platform-const')` or `@pragma('vm:platform-const-if, ...)`, and
* the field initializer or body of the static getter can be evaluated to a constant using the additional compile-time information provided (the target OS).

Note that 

* `Platform.operatingSystem`,
* `Platform.isX` for each of the supported platforms,
* `Platform.pathSeparator`, and
* Flutter's [`defaultTargetPlatform`](https://api.flutter.dev/flutter/foundation/defaultTargetPlatform.html)

are all appropriately annotated and thus can be used to create new platform-constant static fields or getters.

Specifying the target OS is done by the build infrastructure automatically when compiling AOT programs, that is, Dart standalone programs (via `dart compile exe`) or AOT snapshots (via `dart compile aot-snapshot`) or building Flutter apps in profile or release mode. That means as a Dart or Flutter programmer, the only thing you need to do is use the existing fields and getters above and appropriately annotate any new static fields or getters you create that are used to conditionally gate platform-specific code and you should reap the tree-shaking benefits 😄
